core/src/rust/filodb_core/src/query/filodb_query.rs

```diff
@@ -0,0 +1,265 @@
+
+/// A query that can potentially be cached
+///
+/// We can't just hold a reference to Tantivy's Query object because
+/// they don't implement Hash/Equals so they can't be a key
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum FiloDBQuery {
+    /// A complex query that is serialized in byte form
+    Complex(Arc<Box<[u8]>>),
+    /// Search by part key
+    ByPartKey(Arc<Box<[u8]>>),
+    /// Search by list of part IDs
+    ByPartIds(Arc<Box<[i32]>>),
+    /// Search by end time
+    ByEndTime(i64),
+    /// Search for single part ID (not cached)
+    ByPartId(i32),
+    /// All docs query (not cached)
+    All,
+}
+
+impl tantivy_utils::query::cache::CachableQuery for FiloDBQuery {
+    fn should_cache(&self) -> bool {
+        match self {
+            FiloDBQuery::Complex(_) => true,
+            FiloDBQuery::ByPartIds(_) => true,
+            FiloDBQuery::ByEndTime(_) => true,
+            // No point caching all docs - the "query" is constant time anyway
+            &FiloDBQuery::All => false,
+            // A single term lookup is very efficient - no benefit in caching the doc ID
+            FiloDBQuery::ByPartId(_) => false,
+            // Also single term lookup
+            FiloDBQuery::ByPartKey(_) => false,
+        }
+    }
+
+    fn to_query(
+        &self,
+        schema: &Schema,
+        default_field: Option<Field>,
+    ) -> Result<Box<dyn Query>, TantivyError> {
+        match self {
+            FiloDBQuery::Complex(query_bytes) => {
+                let (_, query) = parse_query(query_bytes, schema, default_field)
+                    .map_err(|e| TantivyError::InternalError(format!("{:#}", e)))?;
+
+                Ok(query)
+            }
+            FiloDBQuery::ByPartKey(part_key) => {
+                let field = schema.get_field(field_constants::PART_KEY)?;
+                let term = Term::from_field_bytes(field, part_key);
+                let query = TermQuery::new(term, IndexRecordOption::Basic);
+
+                Ok(Box::new(query))
+            }
+            FiloDBQuery::ByPartIds(part_ids) => {
+                let part_id_field = schema.get_field(field_constants::PART_ID)?;
+
+                let mut terms = vec![];
+                for id in part_ids.iter() {
+                    let term = Term::from_field_i64(part_id_field, *id as i64);
+                    terms.push(term);
+                }
+
+                let query = TermSetQuery::new(terms);
+
+                Ok(Box::new(query))
+            }
+            FiloDBQuery::All => Ok(Box::new(AllQuery)),
+            FiloDBQuery::ByPartId(part_id) => {
+                let part_id_field = schema.get_field(field_constants::PART_ID)?;
+                let term = Term::from_field_i64(part_id_field, *part_id as i64);
+
+                let query = TermQuery::new(term, IndexRecordOption::Basic);
+
+                Ok(Box::new(query))
+            }
+            FiloDBQuery::ByEndTime(ended_before) => {
```

Comment:
Nit pick: I know this is how its named in the Lucent implementation, but can the variable be ended_at? ``before`` makes one believe its end Exclusive.

Reply:
Updated